sycl/test-e2e/VirtualFunctions/misc/range-non-uniform-vf.cpp

```diff
@@ -0,0 +1,98 @@
+// REQUIRES: aspect-usm_shared_allocations
+//
+// This test checks that virtual functions work correctly within simple range
+// kernels when different work-items perform a virtual function calls using
+// different objects.
+//
+// RUN: %{build} -o %t.out %helper-includes
+// RUN: %{run} %t.out
+
+#include <sycl/builtins.hpp>
+#include <sycl/detail/core.hpp>
+#include <sycl/usm.hpp>
+
+#include "helpers.hpp"
+
+#include <iostream>
+
+namespace oneapi = sycl::ext::oneapi::experimental;
+
+class BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual float apply(float) = 0;
+};
+
+class FloorOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual float apply(float V) { return sycl::floor(V); }
+};
+
+class CeilOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual float apply(float V) { return sycl::ceil(V); }
+};
+
+class RoundOp : public BaseOp {
+public:
+  SYCL_EXT_ONEAPI_FUNCTION_PROPERTY(oneapi::indirectly_callable)
+  virtual float apply(float V) { return sycl::round(V); }
+};
+
+int main() try {
+  using storage_t = obj_storage_t<FloorOp, CeilOp, RoundOp>;
+
+  std::array<storage_t, 3> HostStorage;
+
+  sycl::queue q;
+
+  auto *DeviceStorage = sycl::malloc_shared<storage_t>(3, q);
+  sycl::range R{1024};
+
+  constexpr oneapi::properties props{oneapi::assume_indirect_calls};
+  {
+    std::vector<float> HostData(R.size());
+    for (size_t I = 1; I < HostData.size(); ++I)
+      HostData[I] = HostData[I - 1] + 0.7;
+    std::vector<float> DeviceData = HostData;
+    sycl::buffer<float> DataStorage(DeviceData.data(), R);
+
+    q.submit([&](sycl::handler &CGH) {
+       CGH.single_task([=]() {
+         DeviceStorage[0].construct</* ret type = */ BaseOp>(0);
+         DeviceStorage[1].construct</* ret type = */ BaseOp>(1);
+         DeviceStorage[2].construct</* ret type = */ BaseOp>(2);
+       });
+     }).wait_and_throw();
+
+    q.submit([&](sycl::handler &CGH) {
+      sycl::accessor DataAcc(DataStorage, CGH, sycl::read_write);
+      CGH.parallel_for(R, props, [=](auto it) {
+        // Select VF that corresponds to this work-item
```

Comment:
```suggestion
        // Select virtual function that corresponds to this work-item
```

although I'm biased here as VF usually means "vector factor" to me.

Reply:
We actually select an object and not a virtual function here, fixed in 21adc2562747149d50c17aa47cc400a3b65dbf5a